src/lsp4clj/server.clj

```diff
@@ -259,14 +197,95 @@
   protocols.endpoint/IEndpoint
   (start [this context]
     ;; Start receiving messages.
-    (let [pipeline (dispatch-input this context input-ch)]
+    (let [;; In order to process some requests and (all) notifications in
+          ;; series, the language server sometimes needs to block
+          ;; client-initiated input. If the language server sends requests
+          ;; during that time, it needs to receive responses, even though it's
+          ;; blocking other input. Otherwise, it will end up in a deadlock,
+          ;; where it's waiting to receive a response off the input-ch and the
+          ;; input-ch isn't being read from because the server is blocking
+          ;; input. See https://github.com/clojure-lsp/clojure-lsp/issues/1500.
+
+          ;; To avoid this problem we processes client-initiated input (client
+          ;; requests and notifications) and server-initiated input (client
+          ;; responses) separately.
+
+          ;; If the server starts blocking waiting for a response, we buffer the
+          ;; client's requests and notifications until the server is prepared to
+          ;; process them.
+
+          ;; However, if too many client requests and notifications arrive
+          ;; before the response, the buffer fills up. In this situation, we
+          ;; abort all the server's pending requests, hoping that this will
+          ;; allow it to process the client's other messages. We do this to
+          ;; prioritize the client's messages over the server's.
+
+          ;; The situation is different in reverse. The client can also start
+          ;; blocking waiting for a server response. In the meantime, the server
+          ;; can send lots of messages. But this is bad behavior on the server's
+          ;; part. So in this scenario we drop the server's earliest requests
+          ;; and notifications.
+
+          ;; We do that by storing them in a sliding buffer. Because of the
+          ;; sliding buffer:
+          ;; * if the language server sends a message, and
+          ;; * if while processing that message the client sends a request which
+          ;;   causes it to block, and
+          ;; * if the language server sends too many other messages between when
+          ;;   the client started blocking and when the language server responds
+          ;;   to the client's request,
+          ;; * then the language server's messages will start being dropped,
+          ;;   starting from the earliest.
```

Comment:
I think this is incorrect.

What you describe sounds like a sliding buffer for messages _sent by the server_, but here we are using a sliding buffer for messages _sent by the client_. The sliding buffer causes the server to drop client responses.

If I understand you correctly you are describing a scenario where the client waits on the server, therefore stops processing messages from the server, causing back pressure on the server, which manifests as the server getting blocked on its output channel. I don't think this can be resolved by the server.

It seems to me that a LSP _client_ would be ill advised to block indefinitely on a server response, as the client provides the user experience and could decide to ignore old server messages.

I think the lsp4clj pipeline should never silently drop messages.

Reply:
**Edit**: I've removed `server-initiated-input-ch`. But still please let me know if you see a reason we'd need it.

Reply:
> I think, though I'd like clarification, that you're saying you don't think we should drop inbound responses.

Yes, this is what I meant 👍 

> So, I think we can just call `protocols.endpoint/receive-response` directly, bypassing any intermediary buffers. (That is, we can remove `server-initiated-input-ch`.) See anything wrong with that?

Looks fine to me! The only thing that comes to mind is that we rely a lot on known implementations of a _protocol_. If we ever use a different implementation, or change the way `receive-notification` is implemented, we need to keep this in mind. Maybe in a v2 we should reconsider using protocols here if we don't expect multiple implementations.

Reply:
> we rely a lot on known implementations of a `protocol`

I'm not sure that's quite right... this code knows how _this_ implementation of the protocol works, but other implementors of the protocol aren't expected to use this code.

We could remove the protocol. It isn't buying us much. That said, `clojure-lsp` has a test helper which implements it. (That implementation acts as a fake client in integration tests. The protocol is designed to represent either side of LSP comms, a client sending and receiving messages from a server, but also a server sending and receiving messages from a client.) And so in a sense the protocol acts as a tiny amount of documentation that both tools, the server and the fake client, have the same responsibilities. It's not a huge benefit, but doesn't hurt either.

Reply:
You mean: The code that makes assumptions on how the protocol is implemented is only used by the implementation itself, so it is not designed to work with other implementations?